src/Datastore/Entity.php

```diff
@@ -0,0 +1,333 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Datastore;
+
+use ArrayAccess;
+use InvalidArgumentException;
+use JsonSerializable;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * A Datastore Entity
+ *
+ * Entity implements PHP's [ArrayAccess](http://php.net/arrayaccess), allowing
+ * access via the array syntax (example below).
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\ServiceBuilder;
+ *
+ * $cloud = new ServiceBuilder;
+ * $datastore = $cloud->datastore();
+ *
+ * $key = $datastore->key('Person', 'Bob');
+ * $entity = $datastore->entity($key), [
+ *     'firstName' => 'Bob',
+ *     'lastName' => 'Testguy'
+ * ]);
+ *
+ * echo $entity['firstName']; // 'Bob'
+ * $entity['location'] = 'Detroit, MI';
+ * ```
+ */
+class Entity implements JsonSerializable, ArrayAccess
+{
+    use DatastoreTrait;
+
+    /**
+     * @var Key
+     */
+    private $key;
+
+    /**
+     * @var array
+     */
+    private $entity;
+
+    /**
+     * @var array
+     */
+    private $options;
+
+    /**
+     * @param Key $key The Entity's Key, defining its unique identifier
+     * @param array $entity The entity body
+     * @param array $options {
+     *     Configuration Options
+     *
+     *     @type bool $encode If set to true, resources will be
+     *           base64 encoded. `true` by default.
+     *     @type string $cursor Set only when the entity is obtained by a query
+     *           result. If set, the entity cursor can be retrieved from
+     *           {@see Google\Cloud\Datastore\Entity::cursor()}.
+     *     @type string $baseVersion Set only when the entity is obtained by a
+     *           query result. If set, the entity cursor can be retrieved from
+     *           {@see Google\Cloud\Datastore\Entity::baseVersion()}.
+     *     @type array $excludeFromIndexes A list of entity keys to exclude from
+     *           datastore indexes.
+     * }
+     */
+    public function __construct(Key $key, array $entity = [], array $options = [])
+    {
+        $this->key = $key;
+        $this->entity = $entity;
+        $this->options = $options + [
+            'encode' => true,
+            'cursor' => null,
+            'baseVersion' => null,
+            'populatedByService' => false,
+            'excludeFromIndexes' => []
+        ];
+    }
+
+    /**
+     * Get the entity data
+     *
+     * Example:
+     * ```
+     * $data = $entity->get();
+     * ```
+     *
+     * @return array
+     */
+    public function get()
+    {
+        return $this->entity;
+    }
+
+    /**
+     * Set the entity data
+     *
+     * Calling this method replaces the entire entity body. To add or modify a
+     * single value on the entity, use the array syntax for assignment.
```

Comment:
We should also notate that you can use __set. As a side note, I'm a little concerned it might be confusing that we have so many ways to set/get properties on an entity.